content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
++++
+title= "Introducing Hash Field Expirations"
+date= 2025-09-15 00:00:00
+description= "Support for volatile fields in Valkey hash objects"
+authors= ["ranshid"]
+[extra]
+featured = true
+featured_image = "/assets/media/featured/random-08.webp"
++++
+
+One of the great strengths of Valkey has always been its built-in ability to expire keys.
+This simple but powerful mechanism lets developers keep their datasets fresh, automatically clear caches, or enforce session lifetimes without additional logic.
+But there has always been one limitation: expiration worked at the level of whole keys.
+If you stored multiple fields in a hash, you could set a TTL for the hash itself, but you couldn’t say “this field should live for 10 seconds while another field should stick around for 10 minutes.”
+
+That limitation forced developers into awkward choices.
+
+Imagine you’re running a feature flag system. You want to store all of a customer’s feature toggles inside one hash:
+
+* Some flags are short-lived experiments, meant to turn off automatically after a few seconds or minutes.
+
+* Others are long-term rollouts, where the toggle might remain valid for days or weeks.
+
+If you only had key-level expirations, you were forced to compromise.
+You could put all flags for a customer into a single hash, but then they would all share one expiration time.
+Or, you could store each flag in its own key so that each could expire independently — but this would explode the number of keys your system has to manage,
+hurting memory efficiency and increasing overhead for operations.
+
+The inability to set different TTLs per field meant developers either built complex cleanup processes outside of Valkey, or gave up flexibility in how expirations were handled.
+
+## Technical Challenges and Design Alternatives
```

Comment:
Maybe Kyle can help, but the text above this point doesn't really lead into the text after this header. This blog is ultimately a deep dive into tradeoffs about implementing HFE.